portalnet/src/overlay/service.rs

```diff
@@ -489,9 +489,20 @@ where
     /// This happens when a query needs to send a request to a node, when a query has completed,
     /// or when a query has timed out.
     async fn query_event_poll<TQuery: Query<NodeId>>(
-        queries: Arc<RwLock<QueryPool<NodeId, TQuery, TContentKey>>>,
+        queries: &mut QueryPool<NodeId, TQuery, TContentKey>,
     ) -> QueryEvent<TQuery, TContentKey> {
-        future::poll_fn(move |_cx| match queries.write().poll() {
+        if let Poll::Ready(result) = Self::generate_query_event_poll(queries) {
+            result
+        } else {
+            // sleep forever
+            std::future::pending().await
+        }
+    }
+
+    fn generate_query_event_poll<TQuery: Query<NodeId>>(
+        queries: &mut QueryPool<NodeId, TQuery, TContentKey>,
+    ) -> Poll<QueryEvent<TQuery, TContentKey>> {
+        match queries.poll() {
```

Comment:
https://github.com/sigp/discv5/blame/682b51e317216b1e2ec3d75b541cdff66d27cf32/src/service.rs#L1547-L1562

All of this RFC/RFN code is originally based from sigp/discv5 implementation

```suggestion
           future::poll_fn(move |_cx| match queries.poll() {
```

Can't we just do this?

Sigp doesn't use arc<mutex>'s either if you view the link I sent

Reply:
Probably? I'll keep working on it. This PR is definitely still a WIP. I just wanted to get it posted early because I remember Milos saying he might work on lock removal, and I want to avoid duplicate work.

Reply:
I just brought it up because

```
            // sleep forever
            std::future::pending().await
``` 

^ Tastes weird, 


and I think it can just be fully avoided by using `future::poll_fn(move |_cx| match queries.poll() {` instead (which is what sigp/discv5 does and they stopped using a lock here 4ish years ago)

Reply:
Oh I totally agree that that first implementation was not ideal. (it is worth checking, but my understanding is that the code is functionally equivalent, though)

The most important thing I wanted to do is see if the full test suite passes when the locks are gone. It looks like going back to the `poll_fn` implementation still works (can't remember what led me away from it at first), so the PR will almost certainly end up with that approach.

Reply:
Sounds good 🚀